### PR TITLE
 0002-Lift-and-shift-matching-algorithms-5170.patch changes.

### PR DESCRIPTION
**Pull request by bitoandy**
### Pull Request checklist
- [ ] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

## Summary by CodeRabbit

- **New Features**
	- Introduced a suite of matching algorithms to improve parsing capabilities, including optimized functions for handling whitespace, brackets, and nested structures.
	- Added a new test suite for comprehensive validation of matching algorithms.

- **Bug Fixes**
	- Removed deprecated methods and classes to streamline functionality and improve performance, addressing potential issues with whitespace and bracket handling.

- **Tests**
	- Enhanced test coverage for the `Ref` class and the new matching algorithms, ensuring robust validation of functionality and handling edge cases.

### CODE REVIEW

**Comment by bitoandy** [comment]
AI Code Review is in progress (usually takes <3 minutes unless it's a very large PR).

File: src/sqlfluff/core/parser/match_algorithms.py [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+            break
+
+        # If we have a valid match, stop looking through segments
+        if best_simple_match:
+            break
+
+    if best_simple_match:
+        return best_simple_match
+    else:
+        return ((), MatchResult.from_unmatched(segments), None)
```
Comment (bitoandy) [inline src/sqlfluff/core/parser/match_algorithms.py:143]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/bitomukesh/swe-sqlfluff/pull/3#issuecomment-2289946639>#06d838</a><span> - 08/14/2024, 09:41 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The function raises a NotImplementedError if a matcher does not have a functioning '.simple()' method. This could cause the entire application to crash unexpectedly if such a matcher is encountered. It would be better to handle this more gracefully, perhaps by logging the error and skipping the problematic matcher.
</div>
<div id="fix">
<b>Fix:</b> Instead of raising a NotImplementedError, log the error and skip the problematic matcher to avoid crashing the application.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -98,11 +98,13 @@
                 "All matchers passed to '.look_ahead_match()' are "
                 "assumed to have a functioning '.simple()' option. "
                 "In a future release it will be compulsory for _all_ "
                 "matchables to implement '.simple()'. Please report "
                 "this as a bug on GitHub along with your current query "
                 f"and dialect.\nProblematic matcher: {matcher}"
             )
 +            continue
         simple_raws, simple_types = simple
         assert simple_raws or simple_types
         if simple_raws:
             if trimmed_seg in simple_raws:
                 simple_match = matcher
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

File: src/sqlfluff/core/parser/match_algorithms.py [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+                            # that bracket, otherwise add them to the buffer
+                            if bracket_stack:
+                                bracket_stack[-1].segments += new_segments
+                            else:
+                                pre_seg_buff += new_segments
+                            seg_buff = match.unmatched_segments
+                            continue
+                        else:
+                            # The types don't match. Error.
+                            raise SQLParseError(
```
Comment (bitoandy) [inline src/sqlfluff/core/parser/match_algorithms.py:277]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/bitomukesh/swe-sqlfluff/pull/3#issuecomment-2289946639>#06d838</a><span> - 08/14/2024, 09:41 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The use of 'RuntimeError' for unexpected states is not ideal. It would be better to use a more specific custom exception to handle unexpected states, which can provide more context and be easier to debug.
</div>
<div id="fix">
<b>Fix:</b> Define a custom exception class for unexpected states and use it instead of 'RuntimeError'. This will make the error handling more specific and provide better context for debugging.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -285,7 +285,7 @@
                     raise RuntimeError("I don't know how we get here?!")
             else:  # pragma: no cover
                 # No match, we're in a bracket stack. Error.
                 raise SQLParseError(
                     "Couldn't find closing bracket for opening bracket.",
                     segment=bracket_stack[-1].bracket,
                 )
         else:
             # No, we're open to more opening brackets or the thing(s)
             # that we're otherwise looking for.
             pre, match, matcher = look_ahead_match(
                 seg_buff,
                 matchers + bracket_matchers,
                 parse_context=parse_context,
             )
 +class UnexpectedStateError(Exception):
 +    """Exception raised for unexpected states in parsing."""
 +
 +    def __init__(self, message="Unexpected state encountered"):
 +        self.message = message
 +        super().__init__(self.message)
 +
 @@ -285,7 +285,7 @@
                     raise UnexpectedStateError("I don't know how we get here?!")
             else:  # pragma: no cover
                 # No match, we're in a bracket stack. Error.
                 raise SQLParseError(
                     "Couldn't find closing bracket for opening bracket.",
                     segment=bracket_stack[-1].bracket,
                 )
         else:
             # No, we're open to more opening brackets or the thing(s)
             # that we're otherwise looking for.
             pre, match, matcher = look_ahead_match(
                 seg_buff,
                 matchers + bracket_matchers,
                 parse_context=parse_context,
             )
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

File: test/core/parser/grammar/grammar_ref_test.py [hunk 3]
```diff
@@ -23,6 +23,12 @@ def test__parser__grammar__ref_eq():
     assert r1 not in check_list
 
 
+def test__parser__grammar__ref_repr():
+    """Test the __repr__ method of Ref."""
+    assert repr(Ref("foo")) == "<Ref: 'foo'>"
+    assert repr(Ref("bar", optional=True)) == "<Ref: 'bar' [opt]>"
```
Comment (bitoandy) [inline test/core/parser/grammar/grammar_ref_test.py:29]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/bitomukesh/swe-sqlfluff/pull/3#issuecomment-2289946639>#06d838</a><span> - 08/14/2024, 09:41 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The new test function 'test__parser__grammar__ref_repr' lacks proper structure and context setup. It should be integrated into a test class and use a testing framework like 'unittest' to ensure proper test discovery and execution.
</div>
<div id="fix">
<b>Fix:</b> Integrate the test function into a test class and use the 'unittest' framework to ensure proper test discovery and execution.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -23,6 +23,12 @@ def test__parser__grammar__ref_eq():
      assert r1 not in check_list
 
 
 +import unittest
 +
 +class TestRef(unittest.TestCase):
 +    def test__parser__grammar__ref_repr(self):
 +        """Test the __repr__ method of Ref."""
 +        self.assertEqual(repr(Ref("foo")), "<Ref: 'foo'>")
 +        self.assertEqual(repr(Ref("bar", optional=True)), "<Ref: 'bar' [opt]>")
 +
  def test__parser__grammar_ref_exclude(generate_test_segments, fresh_ansi_dialect):
      """Test the Ref grammar exclude option."""
      ni = Ref("NakedIdentifierSegment", exclude=Ref.keyword("ABS"))
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

File: src/sqlfluff/core/parser/grammar/greedy.py [hunk 4]
```diff
@@ -243,12 +144,12 @@ def match(
 
         # NB: If all we matched in the greedy match was non-code then we can't
         # claim it.
-        if not any(seg.is_code for seg in greedy_match.matched_segments):
+        if not any(seg.is_code for seg in greedy_matched.matched_segments):
```
Comment (bitoandy) [inline src/sqlfluff/core/parser/grammar/greedy.py:147]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/bitomukesh/swe-sqlfluff/pull/3#issuecomment-2289946639>#06d838</a><span> - 08/14/2024, 09:41 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The variable 'greedy_matched' is used without validation. If 'greedy_matched' is 'None' or does not have the expected attributes ('matched_segments' and 'unmatched_segments'), it will raise an 'AttributeError'.
</div>
<div id="fix">
<b>Fix:</b> Add validation to ensure 'greedy_matched' is not 'None' and has the required attributes before accessing them.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -144,12 +144,16 @@
       # NB: If all we matched in the greedy match was non-code then we can't
       # claim it.
 -     if not any(seg.is_code for seg in greedy_matched.matched_segments):
 +     if greedy_matched is None or not hasattr(greedy_matched, 'matched_segments') or not any(seg.is_code for seg in greedy_matched.matched_segments):
           # So just return the original match.
           return match
      
       # Otherwise Combine the results.
       return MatchResult(
 -         match.matched_segments + greedy_matched.matched_segments,
 -         greedy_matched.unmatched_segments,
 +         match.matched_segments + (greedy_matched.matched_segments if hasattr(greedy_matched, 'matched_segments') else []),
 +         greedy_matched.unmatched_segments if hasattr(greedy_matched, 'unmatched_segments') else [],
       )
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

File: test/core/parser/match_algorithms_test.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    assert matcher == fs
+    # We shouldn't match the whitespace with the keyword
+    assert match.matched_segments == (
+        KeywordSegment("foo", bracket_seg_list[8].pos_marker),
+    )
+    # Check that the unmatched segments are nothing.
+    assert not match.unmatched_segments
+
+
+def test__parser__algorithms__bracket_fail_with_open_paren_close_square_mismatch(
```
Comment (bitoandy) [inline test/core/parser/match_algorithms_test.py:126]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/bitomukesh/swe-sqlfluff/pull/3#issuecomment-2289946639>#06d838</a><span> - 08/14/2024, 09:41 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The test case for mismatched brackets is not comprehensive enough. It only tests one specific scenario. There should be additional test cases to cover other types of mismatched brackets and edge cases.
</div>
<div id="fix">
<b>Fix:</b> Add additional test cases to cover other types of mismatched brackets and edge cases to ensure comprehensive testing.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -156,6 +156,26 @@
     )
   assert sql_parse_error.match("Found unexpected end bracket")

 +def test__parser__algorithms__bracket_fail_with_open_square_close_paren_mismatch(
 +    generate_test_segments, fresh_ansi_dialect
 +):
 +    """Test bracket_sensitive_look_ahead_match failure case.
 +
 +    Should fail when the type of a close bracket doesn't match the type of the
 +    corresponding open bracket, but both are "definite" brackets.
 +    """
 +    fs = StringParser("foo", KeywordSegment)
 +    # We need a dialect here to do bracket matching
 +    ctx = ParseContext(dialect=fresh_ansi_dialect)
 +    with pytest.raises(SQLParseError) as sql_parse_error:
 +        bracket_sensitive_look_ahead_match(
 +            generate_test_segments(
 +                [
 +                    "select",
 +                    " ",
 +                    "*",
 +                    " ",
 +                    "from",
 +                    "[",
 +                    "foo",
 +                    ")",  # Bracket types don't match (square vs parens)
 +                ]
 +            ),
 +            [fs],
 +            ctx,
 +        )
 +    assert sql_parse_error.match("Found unexpected end bracket")
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

File: test/core/parser/match_algorithms_test.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+def test__parser__algorithms__bracket_fail_with_unexpected_end_bracket(
+    generate_test_segments, fresh_ansi_dialect
+):
+    """Test bracket_sensitive_look_ahead_match edge case.
+
+    Should fail gracefully and stop matching if we find a trailing unmatched.
+    """
+    fs = StringParser("foo", KeywordSegment)
+    # We need a dialect here to do bracket matching
+    ctx = ParseContext(dialect=fresh_ansi_dialect)
```
Comment (bitoandy) [inline test/core/parser/match_algorithms_test.py:167]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/bitomukesh/swe-sqlfluff/pull/3#issuecomment-2289946639>#06d838</a><span> - 08/14/2024, 09:41 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The test case does not include any setup or teardown methods, which are essential for ensuring that the test environment is correctly initialized and cleaned up after the test runs. This can lead to flaky tests and make debugging difficult.
</div>
<div id="fix">
<b>Fix:</b> Add setup and teardown methods to ensure that the test environment is correctly initialized and cleaned up after the test runs.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -0,0 +167,26 @@
 +    def setUp(self):
 +        self.ctx = ParseContext(dialect=fresh_ansi_dialect)
 +        self.segments = generate_test_segments(
 +            [
 +                "bar",
 +                "(",  # This bracket pair should be mutated
 +                ")",
 +                " ",
 +                ")",  # This is the unmatched bracket
 +                " ",
 +                "foo",
 +            ]
 +        )
 +        self.matchers = [fs]
 +
 +    def tearDown(self):
 +        self.ctx = None
 +        self.segments = None
 +        self.matchers = None
 +
 +    def test_bracket_sensitive_look_ahead_match(self):
 +        _, match, _ = bracket_sensitive_look_ahead_match(
 +            self.segments,
 +            self.matchers,
 +            self.ctx,
 +        )
 +        # Check we don't match (even though there's a foo at the end)
 +        assert not match
 +        # Check the first bracket pair have been mutated.
 +        segs = match.unmatched_segments
 +        assert segs[1].is_type("bracketed")
 +        assert segs[1].raw == "()"
 +        assert len(segs[1].segments) == 2
 +        # Check the trailing foo hasn't been mutated
 +        assert segs[5].raw == "foo"
 +        assert not isinstance(segs[5], KeywordSegment)
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

File: src/sqlfluff/core/parser/match_algorithms.py (hunk 1 above)
Comment (bitoandy) [reply src/sqlfluff/core/parser/match_algorithms.py:143]: good to have

File: src/sqlfluff/core/parser/match_algorithms.py (hunk 2 above)
Comment (bitoandy) [reply src/sqlfluff/core/parser/match_algorithms.py:277]: Good to have, not necessary but a good practice.
Our code suggestion could have been better. + and - in some cases are missing in lines.

File: test/core/parser/grammar/grammar_ref_test.py (hunk 3 above)
Comment (bitoandy) [reply test/core/parser/grammar/grammar_ref_test.py:29]: this is invalid suggestion as pytest is being used here. Need to discover the test framework

File: src/sqlfluff/core/parser/grammar/greedy.py (hunk 4 above)
Comment (bitoandy) [reply src/sqlfluff/core/parser/grammar/greedy.py:147]: Good to have, not necessary but a good practice

File: test/core/parser/match_algorithms_test.py (hunk 5 above)
Comment (bitoandy) [reply test/core/parser/match_algorithms_test.py:126]: The issue and fix mentioned are correct,.
BUT the code suggestion is incorrect as it replicates the code in diff without any change.
Either we should have provided additional test cases in code suggestion or skipped the code suggestion altogether.
Although the issue and fix are valid **the code suggestion makes it invalid**
